state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1246,154 +1220,248 @@ pub mod pallet {
+						_ => {
+							if dca_state.scheduled_chunk_swap_id != Some(swap.swap_id()) {
+								log_or_panic!(
+									"Executed swap with unexpected id {} for swap request {swap_request_id}", swap.swap_id()
 								);
-							},
+							}
+
+							false
 						},
-					SwapRequestState::Regular { output_address } => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
-						match T::EgressHandler::schedule_egress(
-							swap.output_asset(),
-							output_amount,
-							output_address.clone(),
-							None, /* ccm metadata */
-						) {
-							Ok(ScheduledEgressDetails {
-								egress_id,
-								egress_amount,
-								fee_withheld,
-							}) => {
-								Self::deposit_event(Event::<T>::SwapEgressScheduled {
-									swap_request_id: swap.swap.swap_request_id,
-									egress_id,
-									asset: swap.output_asset(),
-									amount: egress_amount,
-									fee: fee_withheld,
-								});
-							},
-							Err(err) => {
-								Self::deposit_event(Event::<T>::SwapEgressIgnored {
-									swap_request_id,
-									asset: swap.output_asset(),
-									amount: output_amount,
-									reason: err.into(),
-								});
-							},
-						};
-					},
-					SwapRequestState::NetworkFee => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
-						if swap.output_asset() == Asset::Flip {
-							FlipToBurn::<T>::mutate(|total| {
-								total.saturating_accrue(output_amount);
-							});
-						} else {
-							log_or_panic!(
-								"NetworkFee burning should not be in asset: {:?}",
-								swap.output_asset()
-							);
-						}
-					},
-					SwapRequestState::IngressEgressFee => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
+					};
 
-						if swap.output_asset() ==
-							ForeignChain::from(swap.output_asset()).gas_asset()
-						{
-							T::IngressEgressFeeHandler::accrue_withheld_fee(
-								swap.output_asset(),
-								output_amount,
+					if is_gas_swap {
+						*gas_swap_state = GasSwapState::OutputReady { gas_budget: output_amount };
+					} else {
+						// The executed swap must be for the principal amount:
+						dca_state.accumulated_output_amount += output_amount;
+
+						// Schedule any remaining DCA chunks:
+						dca_state.scheduled_chunk_swap_id = if dca_state.remaining_chunks > 0 {
+							let chunk_input_amount = dca_state.prepare_next_chunk();
+
+							let swap_id = Self::schedule_swap(
+								request.input_asset,
+								request.output_asset,
+								chunk_input_amount,
+								swap.swap.refund_params.clone(),
+								SwapType::CcmPrincipal,
+								swap_request_id,
 							);
+
+							Some(swap_id)
 						} else {
-							log_or_panic!(
-								"IngressEgressFee swap should not be to non-gas asset: {:?}",
-								swap.output_asset()
+							None
+						};
+
+						// See if we still need to schedule the gas swap
+						if let GasSwapState::ToBeScheduled { gas_budget, other_gas_asset } =
```

Comment:
Does this mean the gas swap is scheduled at the same block as the last chunk of the dca swap? Is this intentional?

Reply:
No, it is scheduled at the same time as the second chunk (i.e. as soon as we have one successful chunk).

Reply:
Ok - I thought the idea was to fully execute the principal swap and then do the gas swap. But I suppose this makes sense.